sdm/src/main/java/com/sap/cds/sdm/handler/TokenHandler.java

```
@@ -140,6 +158,107 @@ public static String getDITokenUsingAuthorities(
+    Map<String, String> parameters = new HashMap<>();
+    parameters.put("assertion", token);
+    return parameters;
+  }
+
+  public static String generateDITokenFromTokenExchange(
+      String token, SDMCredentials sdmCredentials, JsonObject payloadObj)
+      throws OAuth2ServiceException {
+    String cachedToken = null;
+    CloseableHttpClient httpClient = null;
+    try {
+      httpClient = HttpClients.createDefault();
+      if (sdmCredentials.getClientId() == null) {
+        throw new IOException("No SDM binding found");
+      }
+      Map<String, String> parameters = fillTokenExchangeBody(token, sdmCredentials);
+      HttpPost httpPost =
+          new HttpPost(
+              sdmCredentials.getBaseTokenUrl()
+                  + "/oauth/token?grant_type=urn:ietf:params:oauth:grant-type:jwt-bearer");
+      httpPost.setHeader(HttpHeaders.ACCEPT, MediaType.APPLICATION_JSON.value());
+      httpPost.setHeader(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_FORM_URLENCODED.value());
+      httpPost.setHeader("X-zid", getTokenFields(token).get("zid").getAsString());
+
+      String encoded =
+          java.util.Base64.getEncoder()
+              .encodeToString(
+                  (sdmCredentials.getClientId() + ":" + sdmCredentials.getClientSecret())
+                      .getBytes());
+      httpPost.setHeader("Authorization", "Basic " + encoded);
+
+      List<BasicNameValuePair> basicNameValuePairs =
+          parameters.entrySet().stream()
+              .map(entry -> new BasicNameValuePair(entry.getKey(), entry.getValue()))
+              .collect(Collectors.toList());
+      httpPost.setEntity(new UrlEncodedFormEntity(basicNameValuePairs));
+
+      HttpResponse response = httpClient.execute(httpPost);
+      String responseBody = extractResponseBodyAsString(response);
+      if (response.getStatusLine().getStatusCode() != HttpStatus.SC_OK) {
+        logger.error("Error fetching token with JWT bearer : " + responseBody);
```

In case status is not 200 do we still need to continue the processing? Or can we throw an exception from here and exit?

---

Done